test/torch_optimizer_test.py

```diff
@@ -0,0 +1,525 @@
+
+# Note this end-to-end only covers the BlueFog optimizers for a single machine,
+# all the hierarchical cases are not fully under test yet.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import pytest
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.utils.data import Dataset, DataLoader
+
+import bluefog.torch as bf
+from bluefog.common import topology_util
+
+TEST_ON_GPU = torch.cuda.is_available()
+
+# A linear model for testing
+class LinearNet(nn.Module):
+    def __init__(self, input_dim, output_dim):
+        super(LinearNet, self).__init__()
+        self.fc = nn.Linear(input_dim, output_dim)
+
+    def forward(self, x):
+        return self.fc(x)
+
+# A Simple dataset for testing.
+class SimpleDataset:
+    def __init__(self, x, y):
+        self._x = x
+        self._y = y
+    
+    def __len__(self):
+        return len(self._y)
+    
+    def __getitem__(self, idx):
+        return (torch.tensor(self._x[idx],dtype=torch.float32),
+                torch.tensor(self._y[idx],dtype=torch.float32))
+
+# A ProblemBuilder for the linear problem with a specified input and output dimension.
+# The matrix A are randomly generated now.
+#   y = AX + e, e ~ N(0, noise_level^2)
+class LinearProblemBuilder:
+    def __init__(self, input_dim = 16, output_dim = 3, noise_level = 1e-5):
+        self._input_dim = input_dim
+        self._output_dim = output_dim
+        self._noise_level = noise_level
+        self._matrix_gen_seed = 0
+        self._generate_matrices()
+
+    def _generate_matrices(self):
+        state = np.random.get_state()
+        np.random.seed(self._matrix_gen_seed)
+        self._A = np.random.randn(self._input_dim, self._output_dim)
+        #self._A = np.ones((self._input_dim, self._output_dim))
+        np.random.set_state(state)
+
+    @property
+    def input_dim(self):
+        return self._input_dim
+    
+    @input_dim.setter
+    def input_dim(self, value):
+        if not isinstance(value, int) or value <= 0:
+            raise ValueError("Input dimension should be an integer larger than 0.")
+        self._input_dim = value
+        self._generate_matrices()
+
+    @property
+    def output_dim(self):
+        return self._output_dim
+    
+    @output_dim.setter
+    def output_dim(self, value):
+        if not isinstance(value, int) or value <= 0:
+            raise ValueError("Output dimension should be an integer larger than 0.")
+        self._output_dim = value
+        self._generate_matrices()
+
+    @property
+    def noise_level(self):
+        return self._noise_level
+
+    @noise_level.setter
+    def noise_level(self, value):
+        if value < 0:
+            raise ValueError("Noise level should be larger than or equal to 0.")
+        self._noise_level = value
+
+    def get_dataset(self, num_sample):
+        x = np.random.randn(num_sample, self.input_dim)
```

Comment:
NIT: In the style guide, the variable use the lower case for variable. But in the math, the lower case and upper case represents the vector/scalar and matrix. So they are conflicted. In convention, if there is no confusion in the name, we should follow math convention instead of code. So I prefer to call it X, E, and Y.

Reply:
Done.